filter.c

```diff
@@ -1168,6 +1176,135 @@ static int func_avg(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack,
     }
     return 1;
 }
+static int compare_doubles(const void *lhs, const void *rhs)
+{
+    double arg1 = *(const double*) lhs;
+    double arg2 = *(const double*) rhs;
+    if (arg1 < arg2) return -1;
+    if (arg1 > arg2) return 1;
+    return 0;
+}
+static double median(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return (vals[0] + vals[1]) / 2;
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            const int middle = n / 2;
+            if (n % 2 == 1) {
+                return vals[middle];
+            } else {
+                return (vals[middle - 1] + vals[middle]) / 2;
+            }
+        }
+    }
+}
+static int func_median(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack, int nstack)
+{
+    token_t *tok = stack[nstack - 1];
+    rtok->nvalues = 0;
+    if ( !tok->nvalues ) return 1;
+    double *vals = (double*) malloc(tok->nvalues * sizeof(double));
+    int i, n = 0;
+    for (i=0; i<tok->nvalues; i++)
+        if ( !bcf_double_is_missing(tok->values[i]) ) { vals[n] = tok->values[i]; n++; }
+    if ( n )
+    {
+        rtok->values[0] = median(vals, n);
+        rtok->nvalues   = 1;
+    }
+    free(vals);
+    return 1;
+}
+static double mode(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return vals[0] < vals[1] ? vals[0] : vals[1];
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            // discretize real values
+            int *int_vals = (int*) malloc(n * sizeof(int));
```

Comment:
malloced addresses should be checked for nullness.

Reply:
Checked how? I don't see any checks on other `malloc`s in filter.c.